pandas/tseries/offsets.py

```diff
@@ -261,16 +261,12 @@ def apply_index(self, i):
         # relativedelta/_offset path only valid for base DateOffset
         if (self._use_relativedelta and
             set(self.kwds).issubset(relativedelta_fast)):
+
             months = ((self.kwds.get('years', 0) * 12
                         + self.kwds.get('months', 0)) * self.n)
             if months:
-                base = (i.to_period('M') + months).to_timestamp()
-                time = i.to_perioddelta('D')
-                days = i.to_perioddelta('M') - time
-                # minimum prevents month-end from wrapping
-                day_offset = np.minimum(days,
-                                        to_timedelta(base.days_in_month - 1, unit='D'))
-                i = base + day_offset + time
+                shifted = tslib.shift_months(i.asi8, months)
+                i = i._constructor(shifted)
```

Comment:
It actually doesn't cause a problem in this case because the values are going to be unboxed/boxed here anyways:
https://github.com/pydata/pandas/blob/master/pandas/tseries/index.py#L716

But I'll change it if it's the more idiomatic way to get the constructor?